File: arsceneview/src/main/java/io/github/sceneview/ar/recording/ARRecorder.kt
package io.github.sceneview.ar.recording

import android.content.ContentValues
import android.content.Context
import android.net.Uri
import android.os.Build
import android.os.Environment
import android.provider.MediaStore
import android.util.Log
import android.util.Size
import android.view.Surface
import androidx.annotation.RequiresApi
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import com.google.ar.core.CameraConfig
import com.google.ar.core.CameraConfigFilter
import com.google.ar.core.PlaybackStatus
import com.google.ar.core.RecordingConfig
import com.google.ar.core.Session
import com.google.ar.core.exceptions.RecordingFailedException
import java.io.File
import java.io.FileNotFoundException
import java.util.EnumSet
import java.util.concurrent.atomic.AtomicReference

/**
 * Compose-friendly helper that records the full state of an [ARSceneView][io.github.sceneview.ar.ARSceneView]'s
 * underlying ARCore [Session] into an MP4 file.
 *
 * ### What ARCore actually records
 *
 * `Session.startRecording(RecordingConfig)` captures **everything** the session sees: the
 * camera image stream, IMU data, depth (if enabled), light estimation, plane updates,
 * anchors, augmented images, and the rest. The resulting MP4 is not a normal video — it is a
 * full session dataset that can be replayed 1:1 with [Session.setPlaybackDataset].
 *
 * ### Why this exists
 *
 * Recording / playback is the single most useful debugging tool ARCore offers: capture a real
 * outdoor / on-site session once, then iterate at the desk against the recording. Combined
 * with the [RerunBridge][io.github.sceneview.ar.rerun.RerunBridge], a recording becomes the
 * deterministic input of a Rerun-driven debugging loop. It also makes deterministic AR tests
 * possible: ship a fixture recording with the test suite and rerun it on every CI.
 *
 * ### Usage
 *
 * ```kotlin
 * val recorder = rememberARRecorder()
 * ARSceneView(
 *     onSessionUpdated = { session, frame -> recorder.attach(session) },
 *     content = { /* … */ }
 * )
 * Button(onClick = { recorder.start(File(context.getExternalFilesDir("ar-recordings"), "session.mp4")) }) {
 *     Text("Record")
 * }
 * Button(onClick = { recorder.stop() }) { Text("Stop") }
 * ```
 *
 * The recorder is non-fatal: any [RecordingFailedException] thrown by ARCore is captured into
 * [errorMessage] and surfaced via the [State.ERROR] state. The session itself keeps running.
 *
 * ### Threading
 *
 * `attach`, `start` and `stop` may be called from any thread — the underlying [AtomicReference]
 * makes the session pointer safe to publish across the render thread (where `onSessionUpdated`
 * runs) and the UI thread (where the buttons live). The actual ARCore calls are synchronous;
 * they're cheap (no blocking I/O), so calling from a UI handler is fine.
 *
 * @see com.google.ar.core.Session.startRecording
 * @see com.google.ar.core.Session.stopRecording
 * @see com.google.ar.core.Session.setPlaybackDataset
 */
public class ARRecorder {

    /** Lifecycle states for the recorder. */
    public enum class State {
        /** No recording in progress. */
        IDLE,

        /** ARCore is actively writing frames to [recordingFile]. */
        RECORDING,

        /** The last [start] / [stop] call failed. See [errorMessage] for details. */
        ERROR,
    }

    private val sessionRef = AtomicReference<Session?>(null)

    /**
     * The session's camera config captured immediately before [start] swapped in a
     * higher-resolution one for `recordingResolution`. `null` when no swap happened — [stop]
     * only restores the config if this is non-null, so a [start] without `recordingResolution`
     * (or one where no matching config was found) leaves the session untouched on [stop].
     */
    private val previousCameraConfig = AtomicReference<CameraConfig?>(null)

    private var _state by mutableStateOf(State.IDLE)
    private var _recordingFile by mutableStateOf<File?>(null)
    private var _errorMessage by mutableStateOf<String?>(null)

    /** Current recorder state. Backed by a Compose `MutableState` — recompose-aware. */
    public val state: State
        get() = _state

    /**
     * The file currently being written (when [state] is [State.RECORDING]) or the most
     * recent file written (after [stop]). `null` until [start] succeeds.
     */
    public val recordingFile: File?
        get() = _recordingFile

    /**
     * Human-readable description of the last error, populated when [state] is [State.ERROR].
     * Cleared on the next successful [start].
     */
    public val errorMessage: String?
        get() = _errorMessage

    /**
     * Bind the recorder to an ARCore [Session]. Call this from `onSessionUpdated` (or any
     * place where you have a live session reference) — passing the same session repeatedly
     * is a no-op, so wiring it on every frame is fine.
     *
     * Without an attached session, [start] returns `false` and transitions to [State.ERROR].
     *
     * **Prefer [recordFrame].** This stateful split between `attach` + `start` was the only
     * AR side-channel in the codebase that didn't pass the session in per call —
     * [io.github.sceneview.ar.rerun.RerunBridge] uses the stateless `logFrame(session, frame)`
     * pattern. Mismatch noted in audit #876; the new [recordFrame] / [start] (Session, File)
     * overloads bring `ARRecorder` in line.
     */
    @Deprecated(
        "Prefer the stateless recordFrame(session) — passes the session per call (matches " +
            "RerunBridge.logFrame). attach() will be removed in v5.",
        ReplaceWith("recordFrame(session)"),
    )
    public fun attach(session: Session) {
        sessionRef.set(session)
    }

    /**
     * Stateless equivalent of [attach]: publishes the latest session reference each frame so
     * a subsequent [start] knows which session to record. Pass the same `session` you got from
     * `onSessionUpdated`. Calling this on every frame with the same session is a no-op (it's
     * an `AtomicReference.set`), matching [io.github.sceneview.ar.rerun.RerunBridge.logFrame].
     */
    public fun recordFrame(session: Session) {
        sessionRef.set(session)
    }

    /**
     * Detach the session reference. After this, [start] will fail until [recordFrame] (or the
     * deprecated [attach]) is called again.
     */
    public fun detach() {
        sessionRef.set(null)
    }

    /**
     * Begin recording the attached session into [file]. The parent directory is created if
     * it doesn't already exist.
     *
     * Recording cannot run while the session is in playback mode — if it is, this returns
     * `false` and transitions to [State.ERROR].
     *
     * @param file               Absolute path of the destination MP4. Passed to
     *                           [RecordingConfig.setMp4DatasetFilePath] (no scoped-storage /
     *                           [android.net.Uri] wrapping needed).
     * @param recordingRotation  Optional [android.view.Display] rotation (`Surface.ROTATION_0` …
     *                           `Surface.ROTATION_270`) so the MP4 plays back upright when
     *                           captured in landscape. Pass the current display rotation from
     *                           the calling context (`display.rotation`). The value is converted
     *                           to degrees via [surfaceRotationToDegrees] before being handed to
     *                           [RecordingConfig.setRecordingRotation], which expects degrees
     *                           (`0`/`90`/`180`/`270`) — not the `Surface.ROTATION_*` ordinal
     *                           (`0`/`1`/`2`/`3`). Default `null` keeps ARCore's default of 0.
     * @param recordingResolution Optional target CPU-image resolution for the recording. ARCore
     *                           writes the **CPU image stream** into the MP4, whose default is
     *                           the lowest-resolution config the device exposes (often 640×480 on
     *                           Pixels). Pass e.g. `Size(1920, 1080)` to record at 1080p — the
     *                           recorder picks the supported BACK-facing, 30 FPS [CameraConfig]
     *                           whose pixel count is closest to the request. `null` (default)
     *                           leaves the session's existing camera config untouched. See #1065.
     *
     * Returns `true` if recording started, `false` if no session is attached, the session is
     * in playback mode, or ARCore refused the start. On failure, [state] is set to
     * [State.ERROR] and [errorMessage] holds the reason.
     */
    public fun start(
        file: File,
        recordingRotation: Int? = null,
        recordingResolution: Size? = null,
    ): Boolean {
        val session = sessionRef.get()
        if (session == null) {
            _errorMessage = "no AR session attached — call recordFrame(session) first"
            _state = State.ERROR
            return false
        }
        return startInternal(session, file, recordingRotation, recordingResolution)
    }

    /**
     * Stateless variant: pass the session explicitly. Equivalent to calling
     * [recordFrame] then [start], minus the runtime "no session attached" failure mode.
     * Mirrors the side-channel shape of
     * [io.github.sceneview.ar.rerun.RerunBridge.logFrame] which also takes the session per
     * call. Once a session is published this way, every subsequent [stop] uses the same
     * reference until [detach] is called or a new session is published.
     *
     * @param recordingResolution Optional target CPU-image resolution — see the [start] overload
     *                            above. `null` (default) keeps the session's camera config as-is.
     */
    public fun start(
        session: Session,
        file: File,
        recordingRotation: Int? = null,
        recordingResolution: Size? = null,
    ): Boolean {
        sessionRef.set(session)
        return startInternal(session, file, recordingRotation, recordingResolution)
    }

    private fun startInternal(
        session: Session,
        file: File,
        recordingRotation: Int?,
        recordingResolution: Size?,
    ): Boolean {
        // Recording is only allowed when the session is NOT in playback mode. PlaybackStatus
        // values: NONE (no playback dataset bound — recordable), OK / FINISHED / IO_ERROR
        // (playback active or terminated — never recordable).
        if (session.playbackStatus != PlaybackStatus.NONE) {
            _errorMessage = "Cannot record while session is in playback mode"
            _state = State.ERROR
            return false
        }
        return try {
            file.parentFile?.mkdirs()
            // ARCore writes the CPU image stream into the MP4 — its default is the
            // lowest-resolution config the device exposes (often 640×480 on Pixels, #1065).
            // When the caller asked for a specific resolution, swap in the closest-matching
            // BACK-facing 30 FPS camera config BEFORE startRecording(). ARCore mandates the
            // camera config be applied while the session is configured (which it is here —
            // the session has been created and configured by ARSceneView); applying it just
            // before startRecording() is the valid window.
            recordingResolution?.let { requested ->
                selectCameraConfig(session, requested)?.let { newConfig ->
                    // Remember the config in effect BEFORE the swap so stop() can restore it —
                    // otherwise the higher-res CPU image stream (and its perf cost) silently
                    // persists for the rest of the AR session (#1358). Captured via runCatching
                    // because Session.getCameraConfig is JNI-backed and must never abort the
                    // recording; on failure we just skip the restore.
                    previousCameraConfig.set(runCatching { session.cameraConfig }.getOrNull())
                    session.cameraConfig = newConfig
                }
            }
            val config = RecordingConfig(session)
                .setMp4DatasetFilePath(file.absolutePath)
                .setAutoStopOnPause(true)
            // ARCore's RecordingConfig.setRecordingRotation(int) expects the display rotation in
            // DEGREES (0/90/180/270). Callers pass a Surface.ROTATION_* constant (ordinals
            // 0/1/2/3) — forwarding it verbatim records a 90° capture as 1° and leaves the
            // dataset sideways (#1648). Convert to degrees before the ARCore call.
            recordingRotation?.let { config.setRecordingRotation(surfaceRotationToDegrees(it)) }
            session.startRecording(config)
            _recordingFile = file
            _errorMessage = null
            _state = State.RECORDING
            true
        } catch (e: RecordingFailedException) {
            logWarning("startRecording failed: ${e.message}")
            // The camera-config swap happens before startRecording() — if the start fails,
            // restore the prior config so a failed recording doesn't leak the high-res
            // stream (#1358).
            restoreCameraConfig(session)
            _errorMessage = e.message ?: "RecordingFailedException"
            _state = State.ERROR
            false
        } catch (e: Exception) {
            logWarning("start failed: ${e.message}")
            restoreCameraConfig(session)
            _errorMessage = e.message ?: e::class.java.simpleName
            _state = State.ERROR
            false
        }
    }

    /**
     * Stop the current recording. Returns the file that was written, or `null` if no
     * recording was active. Safe to call when idle — it's a no-op.
     *
     * If [start] swapped the session's camera config to satisfy a `recordingResolution`
     * request, the prior config is restored here so the higher-res CPU image stream does not
     * outlive the recording (#1358).
     */
    public fun stop(): File? {
        val session = sessionRef.get() ?: run {
            _state = State.IDLE
            return _recordingFile
        }
        return try {
            session.stopRecording()
            _state = State.IDLE
            _recordingFile
        } catch (e: RecordingFailedException) {
            logWarning("stopRecording failed: ${e.message}")
            _errorMessage = e.message ?: "RecordingFailedException"
            _state = State.ERROR
            null
        } catch (e: Exception) {
            logWarning("stop failed: ${e.message}")
            _errorMessage = e.message ?: e::class.java.simpleName
            _state = State.ERROR
            null
        } finally {
            // Restore the camera config even if stopRecording() threw — leaking the
            // higher-res CPU image stream for the rest of the session is worse than the
            // stop failure itself (#1358).
            restoreCameraConfig(session)
        }
    }

    /**
     * Restore the camera config captured by [start] before it swapped in a higher-resolution
     * one for `recordingResolution`. No-op when no swap happened (the [AtomicReference] holds
     * `null`). The swapped-out config is cleared so a subsequent recording without
     * `recordingResolution` does not erroneously restore a stale config.
     *
     * Wrapped in `runCatching` because [Session.setCameraConfig] is JNI-backed; a failure to
     * restore must not surface as a recording error — the recording has already stopped.
     */
    private fun restoreCameraConfig(session: Session) {
        previousCameraConfig.getAndSet(null)?.let { prior ->
            runCatching { session.cameraConfig = prior }
                .onFailure { logWarning("restoring camera config failed: ${it.message}") }
        }
    }

    /** Reset the recorder to [State.IDLE], clearing any previous error. */
    public fun clearError() {
        if (_state == State.ERROR) {
            _state = State.IDLE
            _errorMessage = null
        }
    }

    /**
     * Query the [session] for its supported BACK-facing, 30 FPS camera configs and pick the
     * one whose CPU image size best matches [requested]. Returns `null` if ARCore exposes no
     * matching config (degenerate device) so the caller leaves the existing config untouched.
     *
     * Wrapped in `runCatching` because [Session.getSupportedCameraConfigs] is JNI-backed and
     * can throw on a session that is not in a queryable state — a failure here must never
     * abort the recording, it just falls back to ARCore's default config.
     */
    private fun selectCameraConfig(session: Session, requested: Size): CameraConfig? =
        runCatching {
            val filter = CameraConfigFilter(session)
                .setFacingDirection(CameraConfig.FacingDirection.BACK)
                .setTargetFps(EnumSet.of(CameraConfig.TargetFps.TARGET_FPS_30))
            pickCameraConfig(session.getSupportedCameraConfigs(filter), requested)
        }.getOrNull()

    private fun logWarning(msg: String) {
        try { Log.w(TAG, msg) } catch (_: RuntimeException) { /* unit test stub */ }
    }

    public companion object {
        private const val TAG = "ARRecorder"

        /**
         * Convert an `android.view.Display` rotation — a [Surface.ROTATION_0] … [Surface.ROTATION_270]
         * constant whose underlying ordinals are `0`/`1`/`2`/`3` — to the **degrees** value
         * (`0`/`90`/`180`/`270`) expected by ARCore's [RecordingConfig.setRecordingRotation].
         *
         * Passing the `Surface.ROTATION_*` ordinal straight through records a 90° physical
         * rotation as `1°`, so the dataset plays back sideways (#1648). Any unrecognised value
         * falls back to `0` — the same default ARCore uses when no rotation is set.
         *
         * Pure function — no ARCore session interaction — so the mapping is unit-testable.
         */
        @JvmStatic
        public fun surfaceRotationToDegrees(surfaceRotation: Int): Int = when (surfaceRotation) {
            Surface.ROTATION_90 -> 90
            Surface.ROTATION_180 -> 180
            Surface.ROTATION_270 -> 270
            else -> 0
        }

        /**
         * Pick the [CameraConfig] from [configs] whose CPU [image size][CameraConfig.getImageSize]
         * is the closest match to [requested] (smallest absolute pixel-count difference). Ties are
         * broken towards the higher-resolution config, and a config exactly matching [requested]
         * always wins.
         *
         * Pure function — no ARCore session interaction — so the selection logic is unit-testable
         * against a mocked [Session.getSupportedCameraConfigs] result. Returns `null` for an empty
         * list so callers can leave the session's camera config untouched on a degenerate device.
         *
         * @see selectCameraConfig
         */
        @JvmStatic
        public fun pickCameraConfig(configs: List<CameraConfig>, requested: Size): CameraConfig? {
            if (configs.isEmpty()) return null
            val targetPixels = requested.width.toLong() * requested.height.toLong()
            return configs.minWithOrNull(
                compareBy<CameraConfig> { config ->
                    val size = config.imageSize
                    val pixels = size.width.toLong() * size.height.toLong()
                    kotlin.math.abs(pixels - targetPixels)
                }.thenByDescending { config ->
                    val size = config.imageSize
                    size.width.toLong() * size.height.toLong()
                },
            )
        }

        /**
         * Copy a recorded MP4 from app-private storage into the public **Downloads** collection
         * so it can be pulled via `adb pull /sdcard/Download/SceneView/<name>.mp4`, opened in
         * the Files app, or shared to Drive / cloud storage.
         *
         * Without this helper, recordings live in [Context.getExternalFilesDir] which is
         * sandboxed to the app — invisible to ADB and not selectable from any picker. The
         * canonical use case is the deterministic-AR-tests workflow: record a real session
         * once on a device, export it here, ship the MP4 as a fixture in
         * `src/androidTest/assets/ar-recordings/`, and replay it on every CI run via
         * [io.github.sceneview.ar.ARSceneView]'s `playbackDataset` parameter — no device with
         * AR hardware required for the regression run.
         *
         * On Android 10+ (API 29) uses [MediaStore.Downloads] and needs no runtime permission
         * for files this app authors. On older versions it falls back to the legacy public
         * Downloads directory; that path requires `WRITE_EXTERNAL_STORAGE` if your `minSdk` is
         * below 29, otherwise it relies on the legacy storage scope already granted.
         *
         * @param context      Any [Context] (application or activity).
         * @param recording    The MP4 file to export. Typically the value returned by [stop]
         *                     or [recordingFile].
         * @param displayName  Optional display name. Defaults to `recording.name`.
         * @param subdirectory Subfolder under `Downloads/`. Defaults to `"SceneView"`. Pass an
         *                     empty string to land directly in `Downloads/`.
         * @return The public content [Uri] (Q+) or `file:` Uri (P-) on success, or `null` if
         *         the copy failed.
         * @throws FileNotFoundException if [recording] does not exist.
         */
        @JvmStatic
        @JvmOverloads
        public fun exportToDownloads(
            context: Context,
            recording: File,
            displayName: String = recording.name,
            subdirectory: String = "SceneView",
        ): Uri? {
            if (!recording.isFile) {
                throw FileNotFoundException("Recording does not exist: ${recording.absolutePath}")
            }
            return try {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                    exportViaMediaStore(context, recording, displayName, subdirectory)
                } else {
                    exportLegacy(recording, displayName, subdirectory)
                }
            } catch (e: Exception) {
                try { Log.w(TAG, "exportToDownloads failed: ${e.message}") } catch (_: RuntimeException) {}
                null
            }
        }

        @RequiresApi(Build.VERSION_CODES.Q)
        private fun exportViaMediaStore(
            context: Context, src: File, name: String, subdirectory: String,
        ): Uri? {
            val resolver = context.contentResolver
            val relativePath = if (subdirectory.isBlank()) {
                Environment.DIRECTORY_DOWNLOADS
            } else {
                "${Environment.DIRECTORY_DOWNLOADS}/$subdirectory"
            }
            val values = ContentValues().apply {
                put(MediaStore.Downloads.DISPLAY_NAME, name)
                put(MediaStore.Downloads.MIME_TYPE, "video/mp4")
                put(MediaStore.Downloads.RELATIVE_PATH, relativePath)
                // IS_PENDING gates visibility until we close the OutputStream below — prevents
                // partial reads if the user opens the file picker mid-copy.
                put(MediaStore.Downloads.IS_PENDING, 1)
            }
            val uri = resolver.insert(MediaStore.Downloads.EXTERNAL_CONTENT_URI, values)
                ?: return null
            try {
                resolver.openOutputStream(uri)?.use { out ->
                    src.inputStream().use { it.copyTo(out) }
                } ?: run {
                    resolver.delete(uri, null, null)
                    return null
                }
                values.clear()
                values.put(MediaStore.Downloads.IS_PENDING, 0)
                resolver.update(uri, values, null, null)
                return uri
            } catch (e: Exception) {
                resolver.delete(uri, null, null)
                throw e
            }
        }

        @Suppress("DEPRECATION")
        private fun exportLegacy(src: File, name: String, subdirectory: String): Uri {
            val downloads = Environment.getExternalStoragePublicDirectory(
                Environment.DIRECTORY_DOWNLOADS,
            )
            val target = if (subdirectory.isBlank()) downloads else File(downloads, subdirectory)
            target.mkdirs()
            val dest = File(target, name)
            src.copyTo(dest, overwrite = true)
            return Uri.fromFile(dest)
        }
    }
}

/**
 * Creates and remembers an [ARRecorder] tied to the composable lifecycle.
 *
 * If the recorder is still recording when the composable leaves the tree, [ARRecorder.stop]
 * is invoked automatically so the MP4 is finalized cleanly. Without this hook, an
 * unfinished recording would be left in an undefined state on the disk.
 */
@Composable
public fun rememberARRecorder(): ARRecorder {
    val recorder = remember { ARRecorder() }
    DisposableEffect(recorder) {
        onDispose {
            if (recorder.state == ARRecorder.State.RECORDING) {
                recorder.stop()
            }
            recorder.detach()
        }
    }
    return recorder
}

File: SceneViewSwift/Sources/SceneViewSwift/AR/ARRecorder.swift
#if os(iOS)
import Foundation
import ReplayKit
import SwiftUI

/// Records the screen (and optionally microphone) during an AR session via
/// ReplayKit's `RPScreenRecorder`. iOS port of Android's
/// `io.github.sceneview.ar.recording.ARRecorder` — record-only because
/// ARKit, unlike ARCore, does not expose a deterministic playback dataset.
///
/// ### What this records
///
/// `RPScreenRecorder.shared()` captures the screen pixels at native
/// resolution into an MP4. It is NOT an ARSession dataset (no IMU, no
/// depth, no per-frame anchors) — the resulting clip is replayable in
/// the Photos app but cannot be fed back into `ARSession` for
/// deterministic replay. See `docs/docs/cheatsheet-ios.md` parity table
/// (#1036) for the rationale.
///
/// ### Usage
///
/// ```swift
/// @StateObject var recorder = ARRecorder()
///
/// ARSceneView(/* ... */)
/// HStack {
///     if recorder.isRecording {
///         Button("Stop") {
///             Task { try await recorder.stopRecording() }
///         }
///     } else {
///         Button("Record") {
///             Task { try await recorder.startRecording() }
///         }
///     }
/// }
/// ```
///
/// ### Threading
///
/// `startRecording` and `stopRecording` are `@MainActor async` because
/// ReplayKit's bridge requires the recorder to be mutated on the main
/// thread. `isRecording` is `nonisolated` so SwiftUI can read it from
/// any context.
///
/// ### Lineage
///
/// - `arsceneview/src/main/java/io/github/sceneview/ar/recording/ARRecorder.kt`
///   — full Android implementation with record + ARCore replay
/// - GitHub issue [#1032](https://github.com/sceneview/sceneview/issues/1032)
///   — deferred from v4.2.0 iOS parity sprint
/// - Agent 4 of the v4.1.0 5-agent post-ship audit — "drop playback, ship record only"
@MainActor
public final class ARRecorder: ObservableObject {

    /// Lifecycle states for the recorder.
    public enum State: Equatable {
        /// No recording in progress.
        case idle
        /// `RPScreenRecorder` is actively writing frames.
        case recording
        /// The last `startRecording` / `stopRecording` call failed; the
        /// associated value carries the human-readable error message.
        case error(String)
    }

    @Published public private(set) var state: State = .idle
    @Published public private(set) var lastOutputURL: URL? = nil

    /// `true` while ReplayKit reports it is actively recording. Distinct
    /// from `state == .recording` because `state` follows the call
    /// site's intent while this property mirrors the underlying recorder.
    public var isRecording: Bool {
        recorder.isRecording
    }

    /// `true` if ReplayKit is available on this device (always true on
    /// iOS hardware, `false` in the simulator until iOS 17.4+ when
    /// `RPScreenRecorder` finally became simulator-supported for the
    /// non-microphone code paths).
    public var isAvailable: Bool {
        recorder.isAvailable
    }

    private let recorder: RPScreenRecorder

    public init(recorder: RPScreenRecorder = .shared()) {
        self.recorder = recorder
    }

    /// Starts an in-app screen recording. Throws if the recorder is
    /// unavailable (no AR support / no screen-record permission / app
    /// foregrounded a recording-blocking system view).
    ///
    /// Resolves once recording is actually live — the caller can flip a
    /// SwiftUI binding on completion and trust subsequent frames are
    /// being captured.
    @MainActor
    public func startRecording() async throws {
        guard isAvailable else {
            let msg = "RPScreenRecorder reports unavailable — device may not support screen capture, or a previous recording session is still tearing down"
            state = .error(msg)
            throw ARRecorderError.unavailable(msg)
        }
        if recorder.isRecording {
            // Calling startRecording twice is a programmer error; surface
            // it as an Error instead of silently no-opping so the UI can
            // show a "recording already in progress" indicator if needed.
            let msg = "ARRecorder.startRecording called while a recording is already in progress"
            state = .error(msg)
            throw ARRecorderError.alreadyRecording(msg)
        }
        // Bridge ReplayKit's completion-handler API into async/await.
        try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
            recorder.startRecording { [weak self] error in
                Task { @MainActor in
                    guard let self else {
                        continuation.resume()
                        return
                    }
                    if let error {
                        self.state = .error(error.localizedDescription)
                        continuation.resume(throwing: error)
                    } else {
                        self.state = .recording
                        continuation.resume()
                    }
                }
            }
        }
    }

    /// Stops the in-progress recording and writes the captured MP4 to
    /// `outputURL` (defaults to a freshly-generated temp file). Returns
    /// the file URL so the caller can share it / save it to Photos /
    /// upload it.
    ///
    /// - Parameter outputURL: Destination for the MP4. The parent
    ///   directory is created if it does not already exist. Pass `nil`
    ///   to write to `NSTemporaryDirectory()/ARRecording-<uuid>.mov`.
    @MainActor
    @discardableResult
    public func stopRecording(outputURL: URL? = nil) async throws -> URL {
        guard recorder.isRecording else {
            let msg = "ARRecorder.stopRecording called while no recording is in progress"
            state = .error(msg)
            throw ARRecorderError.notRecording(msg)
        }
        let destination = outputURL ?? Self.defaultOutputURL()
        // Ensure the parent directory exists before ReplayKit tries to
        // write into it (RPScreenRecorder will fail if it can't create
        // the file).
        try? FileManager.default.createDirectory(
            at: destination.deletingLastPathComponent(),
            withIntermediateDirectories: true
        )
        return try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<URL, Error>) in
            recorder.stopRecording(withOutput: destination) { [weak self] error in
                Task { @MainActor in
                    guard let self else {
                        if let error {
                            continuation.resume(throwing: error)
                        } else {
                            continuation.resume(returning: destination)
                        }
                        return
                    }
                    if let error {
                        self.state = .error(error.localizedDescription)
                        continuation.resume(throwing: error)
                    } else {
                        self.state = .idle
                        self.lastOutputURL = destination
                        continuation.resume(returning: destination)
                    }
                }
            }
        }
    }

    /// Default destination for a recording when no `outputURL` is
    /// passed to ``stopRecording(outputURL:)``. Lives under
    /// `NSTemporaryDirectory()` so the OS reclaims the file
    /// automatically — the caller is expected to move it to a
    /// persistent location (e.g. via `PHPhotoLibrary`) if they want to
    /// keep it.
    public static func defaultOutputURL() -> URL {
        let filename = "ARRecording-\(UUID().uuidString).mov"
        return URL(fileURLWithPath: NSTemporaryDirectory()).appendingPathComponent(filename)
    }
}

/// Errors surfaced by ``ARRecorder``.
public enum ARRecorderError: LocalizedError, Equatable {
    case unavailable(String)
    case alreadyRecording(String)
    case notRecording(String)

    public var errorDescription: String? {
        switch self {
        case .unavailable(let msg), .alreadyRecording(let msg), .notRecording(let msg):
            return msg
        }
    }
}

/// SwiftUI helper that returns a ``ARRecorder`` instance scoped to the
/// surrounding view's lifetime. Mirrors Android's `rememberARRecorder()`
/// composable — wraps `@StateObject` so the same recorder instance
/// survives recompositions.
///
/// ```swift
/// struct MyARView: View {
///     @StateObject private var recorder = ARRecorder()
///     // ...
/// }
/// ```
///
/// The Android version is a top-level `@Composable fun rememberARRecorder()`;
/// SwiftUI's `@StateObject` property wrapper provides the same lifetime
/// guarantee, so `rememberARRecorder()` ships as a thin convenience
/// initialiser rather than a global function. Use either pattern —
/// `@StateObject` is the idiomatic SwiftUI choice.
public extension ARRecorder {
    /// Convenience initialiser that returns a fresh recorder bound to
    /// the shared `RPScreenRecorder`. Equivalent to
    /// `ARRecorder()` — provided so the call site reads symmetrically
    /// with the Android factory `rememberARRecorder()`.
    static func remembered() -> ARRecorder {
        ARRecorder()
    }
}
#endif // os(iOS)
